### PR TITLE
New voting system for OC

### PR DESCRIPTION
**Pull request by kojoty**
This is the new voting system for OC.

OCPL has voting on OCteam every year and we need new script for this but this script can be use for any simple polling where the options are just flat list etc.

This PR is for the first review of the code I would like to merge later.
I think it's mostly done  I still want to work on the views - csses and views need improvements - hope to update it soon.

Some test voting can be generated by test controller.

### CODE REVIEW

File: src/Models/Voting/ChoiceOption.php [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    {
+        parent::__construct();
+    }
+
+    public static function getOptionsForElection(Election $election): array
+    {
+        $db = self::db();
+        $stmt = $db->multiVariableQuery(
+            "SELECT * FROM vote_choiceOptions WHERE electionId = :1 ORDER BY orderIdx ASC",
+            $election->getElectionId());
```
Comment (rapotek) [inline src/Models/Voting/ChoiceOption.php:43]: I am not sure, but I think the last `);` should be in a new line.

File: src/Models/Voting/ChoiceOption.php [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+     * Returns TRUE if given option is a valid option for given elections
+     *
+     * @param Election $election
+     * @param int $optionId
+     */
+    public static function checkOption(Election $election, int $optionId): bool
+    {
+        return self::db()->multiVariableQueryValue(
+            "SELECT COUNT(*) FROM vote_choiceOptions WHERE optionId = :1 AND electionId = :2",
+            0, $optionId, $election->getElectionId()) > 0;
```
Comment (rapotek) [inline src/Models/Voting/ChoiceOption.php:62]: Also here, I would move `) > 0;` to a new line and ensure each argument of a method call is in a single line.

File: src/Models/Voting/ChoiceOption.php (hunk 1 above)
Comment (kojoty) [reply src/Models/Voting/ChoiceOption.php:43]: of course according to  PSR-2/12 it should be as you wrote - but i don't like this part of PSR :)
(but I will fix it as you so meticulous) 

...and you're right!!! code should be strict!

File: src/Models/Voting/ChoiceOption.php (hunk 2 above)
Comment (kojoty) [reply src/Models/Voting/ChoiceOption.php:62]: same as above (https://github.com/opencaching/opencaching-pl/pull/2216#discussion_r580918226)

File: lib/languages/en.php [hunk 3]
```diff
@@ -3066,4 +3066,38 @@
     'editCache_orderPicsSaveSuccess' => 'Pics order has been changed',
 
 
+    'vote_criteriaNotPassed' => 'Currently logged user account donesn\'t pass criteria of participation in voting',
+    'vote_electionOpenUntil' => 'This election is open until',
+    'vote_electionWillBeOpen' => 'This election will be open',
+    'vote_timezone' => 'Timezone',
+    'vote_saveMyVote' => 'Save my vote',
+    'vote_alreadyVoted' => 'You have voted before and your vote was saved. The results will be showed here when votings ends.',
```
Comment (jrmajor) [inline lib/languages/en.php:3074]:
```suggestion
    'vote_alreadyVoted' => 'You have voted before and your vote was saved. The results will be shown here when votings ends.',
```

File: lib/languages/en.php [hunk 4]
```diff
@@ -3066,4 +3066,38 @@
...
+    'vote_votersCount' => 'A total number of people voted',
+    'vote_votes' => 'Votes',
+    'vote_numOfVotesChartTitle' => '# of votes',
+    'vote_numOfVotesChartAxisY' => '# of votes',
+    'vote_votesInTimeChartTitle' => 'Votes in time',
+    'vote_votesInTimeChartYAxisY' => '# of votes (logarythmic scale)',
+    'vote_resultsSectionTitle' => 'Results of voting',
+    'vote_chartsSectionTitle' => 'Charts',
+    'vote_tooManyOptionsSelected' => 'Too many options selected. Maximum number of allowed options to choose',
+    'vote_noOptionsSelected' => 'Empty vote is not ellowed. Please select something.',
```
Comment (jrmajor) [inline lib/languages/en.php:3092]:
```suggestion
    'vote_noOptionsSelected' => 'Empty vote is not allowed. Please select something.',
```

File: lib/languages/en.php [hunk 5]
```diff
@@ -3066,4 +3066,38 @@
...
+    'vote_notEnoughOptionsSelected' => 'Not enough options selected. Number options to select',
+    'vote_saveResultInternalError' => 'Internal error occured. If problem is persistent please contact administrator.',
+    'vote_saveResultSuccess' => 'Vote successfully saved.',
+    'vote_elListTitle' => 'List of elections',
+    'vote_elListNoElections' => 'There is no elections at all',
```
Comment (jrmajor) [inline lib/languages/en.php:3097]:
```suggestion
    'vote_elListNoElections' => 'There are no elections at all',
```

File: src/Controllers/TestController.php [hunk 6]
```diff
@@ -464,4 +469,104 @@ public function altitudeTest($lat=null, $lon=null)
...
+            }
+
+            $user = User::fromUserIdFactory($userId);
+            if (!$user) {
+                continue;
+            }
+
+            // OK this is next unique user
+            $users[$userId] = $user;
+        } while ($step < 2*$MAX_CANDIDATEDS && count($users) < $candidates);
```
Comment (jrmajor) [inline src/Controllers/TestController.php:521]:
```suggestion
        } while ($step < 2 * $MAX_CANDIDATEDS && count($users) < $candidates);
```

File: src/Controllers/VotingController.php [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        $this->view->setTemplate('voting/election');
+        $this->view->buildView();
+    }
+
+    /**
+     * Save votes to DB.
+     * Votes are in var POST[votes]
+     *
+     * @param int $electionId
+     *
```
Comment (jrmajor) [inline src/Controllers/VotingController.php:88]:
```suggestion
```

File: src/Models/Voting/Election.php [hunk 8]
```diff
@@ -0,0 +1,401 @@
+<?php
+/**
+ * Model of elections - contains information related to election process
+ */
```
Comment (jrmajor) [inline src/Models/Voting/Election.php:4]: I think the docblock should be placed right above the class.

File: src/Models/Voting/ElectionResult.php [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+            $result[] = ['x' => "$ts", 'y' => $currentVotes];
+        }
+        return json_encode(array_values($result));
+    }
+
+    public function getColorForOption(ChoiceOption $opt): string
+    {
+        return $this->optionsColors[$opt->getOptionId()];
+    }
+}
```
Comment (jrmajor) [inline src/Models/Voting/ElectionResult.php:152]: No newline at end of file.

File: src/Utils/Database/OcDb.php [hunk 10]
```diff
@@ -26,6 +28,45 @@ public static function instance($access = self::NORMAL_ACCESS)
...
 
+    /**
+     * Overloading of beginTransaction method to detect nested transations which can't work!
+     * Only one transaction can be active
+     *
+     * {@inheritDoc}
+     * @see OcPdo::beginTransaction()
+     */
+    function beginTransaction () {
+        if ( $this->hasActiveTransaction ) {
```
Comment (jrmajor) [inline src/Utils/Database/OcDb.php:39]:
```suggestion
    public function beginTransaction() {
        if ($this->hasActiveTransaction) {
```

File: src/Utils/Generators/ColorGenerator.php [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+     * @return string
+     */
+    public static function trueRandomColor(): string
+    {
+        return sprintf("#%06X", mt_rand(0, 0xFFFFFF));
+    }
+
+    public static function rgb(string $htmlColorName): array
+    {
+        return self::RGB_HTML_COLORS[$htmlColorName] ?? [0,0,0];
```
Comment (jrmajor) [inline src/Utils/Generators/ColorGenerator.php:356]:
```suggestion
        return self::RGB_HTML_COLORS[$htmlColorName] ?? [0, 0, 0];
```

File: src/Views/chunks/chartsJs.tpl.php [hunk 12]
```diff
@@ -0,0 +1,20 @@
+<?php
+/**
+ * This chunk is used to load chartsJs to main template (main.tpl/mini.tpl etc.),
+ * so there is no need to call it in ordinary content templates.
+ *
+ * The only thing to load chartsJs (this chunk) in template header is to call:
+ *
+ *     $this->view->addHeaderChunk('chartsJs')
+ */
+return function (){
```
Comment (jrmajor) [inline src/Views/chunks/chartsJs.tpl.php:10]:
```suggestion
return function () {
```

File: src/Views/voting/election.tpl.php [hunk 13]
```diff
@@ -0,0 +1,175 @@
+<?php
+use src\Models\Voting\Election;
```
Comment (jrmajor) [inline src/Views/voting/election.tpl.php:2]:
```suggestion
<?php

use src\Models\Voting\Election;
```

File: lib/languages/en.php [hunk 14]
```diff
@@ -3066,4 +3066,38 @@
     'editCache_orderPicsSaveSuccess' => 'Pics order has been changed',
 
 
+    'vote_criteriaNotPassed' => 'Currently logged user account donesn\'t pass criteria of participation in voting',
```
Comment (rapotek) [inline lib/languages/en.php:3069]:
```suggestion
    'vote_criteriaNotPassed' => 'Currently logged user account doesn\'t pass criteria of participation in voting',
```
Just testing how "suggestion" works, I don't remember using it before ;)

**Pull request merged**